### ISSUE TITLE
Add batch normalisation to BCNN

### ISSUE THREAD

**Issue by RustyPanda** [RustyPanda/zoobot#17]
Batch normalization has been shown to make training faster and plateau higher, for deep networks. It prevents gradients from saturating at extreme values by normalizing the layer output according to empirically-observed batch statistics.

For an introduction, see slide 54 here http://cs231n.stanford.edu/slides/2017/cs231n_2017_lecture6.pdf and the original paper 'Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift'.

TensorFlow provides two levels of NN abstraction (three including Keras):
- functional abstraction in 'layers' e.g. https://www.tensorflow.org/api_docs/python/tf/layers/conv2d 
- nuts-and-bolts abstraction in 'nn' e.g. https://www.tensorflow.org/api_docs/python/tf/nn/conv2d
Currently I use the functional abstraction because the other requires explicit filters.

Similarly batch normalisation has
- functional abstraction https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization
- core abstraction https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization
Currently, I add batch norm using the functional abstraction because I don't want to have to track and calculate the batch statistics (mean and variance).

Batch norm should be applied either right before (more common?) or right after the activation function. I have turned off the activation in layers.conv2d, applied batch norm, and then applied the activation.

---

It's not clear which activation is best to use with batch norm. 

In Inception, the batch norm. is be followed by a relu activation.
https://www.tensorflow.org/api_docs/python/tf/nn/relu
It might also be worth trying out a leaky relu or prelu later - see #16

Stanford CS231 shows a diagram using batch norm and tanh activation.

I don't think batch norm applies any activation in tf, but I'm not totally sure - training seems to run okay with NO activation function! https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/layers/normalization.py

**Comment by RustyPanda** [comment]
Currently running with dual preprocessing + batch norm: it's not obviously better than simply with dual preprocessing. I should test this properly with
- dual preprocessing only
- batch norm (without activation?) + dual preprocessing (currently running, 'augs_both_normed')
- batch norm (with activation) + dual preprocessing

**Comment by RustyPanda** [comment]
The batch norm code is a messy copy. If it's clearly better, I should refactor to make default. If not, I will need to ponder how to 
- refactor so both options are available
- record the selected option

**Comment by RustyPanda** [comment]
Using relu activation after batch norm sped up training. 

With dual preprocessing:
- batch norm gets to the same accuracy (1k epochs) with and without activation
- activation avoids the training slowdown from batch norm

**Comment by RustyPanda** [comment]
I need to turn off bias on the conv2d layers. Batch norm will just normalise it away again. Might as well reduce parameters. 

Thanks to this tweet from Karpathy: https://twitter.com/karpathy/status/1013245864570073090?s=19
